File: watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt
package app.vescape.wear

import android.os.SystemClock
import androidx.activity.compose.BackHandler
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.PagerDefaults
import androidx.compose.foundation.pager.VerticalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.PointerEventPass
import androidx.compose.ui.input.pointer.pointerInput
import androidx.wear.compose.foundation.BasicSwipeToDismissBox
import androidx.wear.compose.foundation.edgeSwipeToDismiss
import androidx.wear.compose.foundation.rememberSwipeToDismissBoxState
import androidx.wear.compose.material.MaterialTheme
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Screen shell around the live Watch Frame. The rim gauges ([FrameLayout]) are pinned at the root
 * and drawn once; both pagers are transparent and only swap what sits in the centre of the circle,
 * so the rider never loses speed, duty, battery and temps by swiping. Routes between the Board
 * Move page ([MoveScreen]), the diagnostics pager page ([DiagnosticsScreen]), the
 * waiting/disconnected status layouts and the close prompt. Also owns the refresh tick that ages a
 * stopped stream into DISCONNECTED. The board Lights page ([LightsScreen]) sits between them.
 *
 * The mirror deliberately does not hold the screen on. `FLAG_KEEP_SCREEN_ON` also suppresses
 * ambient, so a ride ran the display at full brightness for hours and the low-power always-on
 * rendering below was unreachable — by far the largest battery cost the wrist had.
 *
 * Ambient does not replace this tree, it settles it: the pagers stay mounted and are parked on the
 * gauges with their gestures off, so going in and out of always-on never rebuilds the screen or
 * restarts the idle clock. What ambient changes is inside [AmbientMode].
 */
@Composable
internal fun MirrorScreen(
    sender: CommandSender,
    ambient: AmbientMode = AmbientOff,
    onRequestClose: () -> Unit = {},
) {
    val isAmbient = ambient.active
    val state by TelemetryState.mirrorState
    val phoneLink by TelemetryState.phoneLink
    var showClosePrompt by remember { mutableStateOf(false) }
    // A hold must not be interpreted as a page swipe, and must never end because the page moved.
    var moveHeld by remember { mutableStateOf(false) }
    // One pager owns the whole vertical axis: weather above the gauges, nav focus below. Two
    // stacked vertical pagers used to compete for the same drag — the inner one claimed the
    // pointer, forwarded the leftover delta to the outer one but kept the velocity, so the outer
    // could only settle by dragging past half the screen and a normal flick sprang back.
    val verticalPagerState = rememberPagerState(initialPage = VERTICAL_PAGE_GAUGES, pageCount = { VERTICAL_PAGE_COUNT })
    // Fractional position on the axis, from radar at 0 up to nav focus at the end. Both focus
    // progresses are read off it, so a drag fades exactly as far as it has travelled.
    val verticalPosition = {
        verticalPagerState.currentPage + verticalPagerState.currentPageOffsetFraction
    }
    // Nav focus is a page of its own so the drag is a real gesture, but nothing new is drawn there:
    // the gauges pin themselves in place (see [navFocus]) and shed their readouts on the way.
    val navFocus = { (verticalPosition() - VERTICAL_PAGE_GAUGES).coerceIn(0f, 1f) }
    val scope = rememberCoroutineScope()
    // Which page owns the vertical axis right now. The radar page fetches while it is on screen and
    // nowhere else, so this is the difference between an idle wrist and a fetching one.
    val onGauges = verticalPagerState.currentPage == VERTICAL_PAGE_GAUGES
    val radarVisible = !isAmbient && verticalPagerState.currentPage == VERTICAL_PAGE_RADAR

    // Mutually exclusive by `enabled`: back closes the innermost thing that is open, and only the
    // gauges themselves treat back as "leave the mirror".
    BackHandler(enabled = showClosePrompt) {
        showClosePrompt = false
    }
    // Anywhere else on the vertical axis returns to the gauges; only the gauges treat back as leave.
    BackHandler(enabled = !showClosePrompt && !onGauges) {
        scope.launch { verticalPagerState.animateScrollToPage(VERTICAL_PAGE_GAUGES) }
    }
    BackHandler(enabled = !showClosePrompt && onGauges) {
        showClosePrompt = true
    }

    LaunchedEffect(isAmbient) {
        while (true) {
            delay(if (isAmbient) AMBIENT_REFRESH_INTERVAL_MS else WATCH_FRAME_INTERVAL_MS)
            TelemetryState.refresh()
        }
    }

    MaterialTheme {
        when {
            showClosePrompt -> Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                ClosePrompt(onStay = { showClosePrompt = false }, onClose = onRequestClose)
            }
            else -> {
                val dismissState = rememberSwipeToDismissBoxState()
                // Page 0 = gauges centre (empty), 1 = Board Move, 2 = board Lights,
                // 3 = diagnostics. Dismiss stays on the left edge; interior horizontal swipes page.
                val controlPagerState = rememberPagerState(pageCount = { CONTROL_PAGE_COUNT })
                var dismissEnabled by remember { mutableStateOf(true) }
                // Idle clock for the auto-return.
                var lastTouchMs by remember { mutableLongStateOf(SystemClock.elapsedRealtime()) }
                // Both pagers are transparent: they only swap the centre of the circle. Their drag
                // offsets are the focus progresses the pinned frame fades its readouts against.
                val controlFocus = {
                    (controlPagerState.currentPage + controlPagerState.currentPageOffsetFraction)
                        .coerceIn(0f, 1f)
                }
                val weatherFocus = { (VERTICAL_PAGE_GAUGES - verticalPosition()).coerceIn(0f, 1f) }
                // The wrist goes always-on wherever the rider left it. Ambient only ever draws the
                // gauges, so park both axes there first — without this the arcs would be pinned
                // over a control page the rider can no longer swipe away.
                LaunchedEffect(isAmbient) {
                    if (!isAmbient) return@LaunchedEffect
                    verticalPagerState.scrollToPage(VERTICAL_PAGE_GAUGES)
                    controlPagerState.scrollToPage(CONTROL_PAGE_GAUGES)
                }
                // A page is interactive only once it has settled: a tap landing mid-swipe belongs
                // to the gesture, not to the control it happened to be over. Derived, so the drag
                // offset does not invalidate the whole mirror on every frame of a swipe.
                val activePage by remember(controlPagerState) {
                    derivedStateOf {
                        controlPagerState.currentPage.takeIf {
                            !controlPagerState.isScrollInProgress &&
                                controlPagerState.currentPageOffsetFraction == 0f
                        }
                    }
                }
                // The readout only answers a tap while the gauges own both axes; mid-swipe or on
                // another page the tap belongs to the page under it.
                val weatherTappable by remember(controlPagerState, verticalPagerState, isAmbient) {
                    derivedStateOf {
                        !isAmbient &&
                            activePage == CONTROL_PAGE_GAUGES &&
                            !verticalPagerState.isScrollInProgress &&
                            verticalPagerState.currentPage == VERTICAL_PAGE_GAUGES
                    }
                }
                // PagerState.settledPage can change before the snap reaches offset zero.
                // Changing the parent modifier then cancels that animation mid-page.
                LaunchedEffect(
                    controlPagerState.isScrollInProgress,
                    verticalPagerState.isScrollInProgress,
                ) {
                    if (!controlPagerState.isScrollInProgress &&
                        controlPagerState.currentPageOffsetFraction == 0f &&
                        !verticalPagerState.isScrollInProgress &&
                        verticalPagerState.currentPageOffsetFraction == 0f
                    ) {
                        dismissEnabled = controlPagerState.currentPage == CONTROL_PAGE_GAUGES &&
                            verticalPagerState.currentPage == VERTICAL_PAGE_GAUGES
                    }
                }
                // Horizontal pages are transient controls, so an untouched wrist drifts back to
                // the gauges. The vertical axis is never moved: weather and the nav-focus map are
                // places a rider parks on deliberately. Re-keying on lastTouchMs restarts the
                // window; a held Move suspends it outright.
                LaunchedEffect(lastTouchMs, moveHeld, isAmbient) {
                    // Ambient has already parked the pager, and an animation there is wasted panel.
                    if (moveHeld || isAmbient) return@LaunchedEffect
                    delay(CONTROL_IDLE_RETURN_MS)
                    if (controlPagerState.currentPage != CONTROL_PAGE_GAUGES) {
                        controlPagerState.animateScrollToPage(CONTROL_PAGE_GAUGES)
                    }
                }
                // The dismiss box wraps the pinned gauges. Outside it, a dismiss drag would slide
                // the transparent pages away while the arcs stayed nailed down.
                BasicSwipeToDismissBox(
                    onDismissed = { showClosePrompt = true },
                    state = dismissState,
                    userSwipeEnabled = dismissEnabled && !isAmbient,
                ) { isBackground ->
                    Box(
                        // Initial pass: the event is seen before any child consumes it, so a tap
                        // that lands on a Move half or does nothing at all still counts as touch.
                        modifier = Modifier
                            .fillMaxSize()
                            .pointerInput(Unit) {
                                awaitPointerEventScope {
                                    while (true) {
                                        awaitPointerEvent(PointerEventPass.Initial)
                                        lastTouchMs = SystemClock.elapsedRealtime()
                                    }
                                }
                            },
                        contentAlignment = Alignment.Center,
                    ) {
                        if (isBackground) {
                            ClosePrompt(onStay = { showClosePrompt = false }, onClose = onRequestClose)
                        } else {
                            VerticalPager(
                                state = verticalPagerState,
                                // The vertical axis belongs to the gauges alone: weather above,
                                // nav focus below. From a control page it would open a blank map
                                // over a page the rider is working on, so it is only live there.
                                userScrollEnabled = !isAmbient && !moveHeld && activePage == CONTROL_PAGE_GAUGES,
                                // A page swap on the wrist is a flick, not a drag: the default
                                // half-screen threshold means a rider has to pull the weather page
                                // most of the way down or watch it spring back.
                                flingBehavior = PagerDefaults.flingBehavior(
                                    state = verticalPagerState,
                                    snapPositionalThreshold = PAGE_SNAP_THRESHOLD,
                                ),
                                modifier = Modifier.fillMaxSize(),
                            ) { verticalPage ->
                                when (verticalPage) {
                                    VERTICAL_PAGE_RADAR -> RadarScreen(visible = radarVisible)
                                    VERTICAL_PAGE_WEATHER -> WeatherScreen()
                                    // Gesture-only page: nothing is drawn here, the drag offset
                                    // alone is the nav-focus progress. Keeping the gauges out of
                                    // the pager is what makes this a transition rather than a page
                                    // swap — a pager clips its pages, so content inside would
                                    // slide away instead of pinning.
                                    VERTICAL_PAGE_NAV -> Unit
                                    else -> HorizontalPager(
                                        state = controlPagerState,
                                        userScrollEnabled = !isAmbient && !moveHeld,
                                        modifier = Modifier
                                            .fillMaxSize()
                                            .then(
                                                if (dismissEnabled) {
                                                    Modifier.edgeSwipeToDismiss(dismissState)
                                                } else {
                                                    Modifier
                                                },
                                            ),
                                    ) { page ->
                                        Box(
                                            modifier = Modifier.fillMaxSize(),
                                            contentAlignment = Alignment.Center,
                                        ) {
                                            when (page) {
                                                // Page 0 is empty: it is the gauges themselves,
                                                // pinned at the root behind this transparent pager.
                                                CONTROL_PAGE_GAUGES -> Unit
                                                CONTROL_PAGE_MOVE -> MoveScreen(
                                                    sender = sender,
                                                    // Ambient parks the pager, but the park is a
                                                    // scroll and this gate is not: a held Move must
                                                    // stop the moment the wrist goes always-on, not
                                                    // once the page has finished travelling.
                                                    interactionEnabled = !isAmbient && activePage == CONTROL_PAGE_MOVE,
                                                    onHoldChanged = { moveHeld = it },
                                                )
                                                CONTROL_PAGE_LIGHTS -> LightsScreen(
                                                    sender = sender,
                                                    interactionEnabled = !isAmbient && activePage == CONTROL_PAGE_LIGHTS,
                                                )
                                                CONTROL_PAGE_DIAGNOSTICS -> DiagnosticsScreen()
                                                else -> Unit
                                            }
                                        }
                                    }
                                }
                            }
                            // Rim arcs, drawn once at the root and never inside a pager: the rider
                            // keeps speed, duty, battery and temps on every page. Drawn *over* the
                            // pagers, because the forecast readout is the one thing here a rider
                            // taps and a full-screen pager above it would eat the tap. Nothing else
                            // in the frame takes pointer input, so the pages stay reachable.
                            MirrorContent(
                                state = state,
                                phoneLink = phoneLink,
                                ambient = ambient,
                                focus = navFocus,
                                controlFocus = controlFocus,
                                weatherFocus = weatherFocus,
                                onWeatherClick = if (weatherTappable) {
                                    {
                                        scope.launch {
                                            verticalPagerState.animateScrollToPage(VERTICAL_PAGE_WEATHER)
                                        }
                                    }
                                } else {
                                    null
                                },
                            )
                        }
                    }
                }
            }
        }
    }
}

/** Fraction of a page a drag must cover to settle on the next one rather than spring back. */
private const val PAGE_SNAP_THRESHOLD = 0.25f

/** Idle time after which the horizontal pager drifts back to the gauges. */
private const val CONTROL_IDLE_RETURN_MS = 45_000L

/**
 * Radar and weather above the gauges, nav focus below: one pager owns the whole vertical axis.
 * Radar sits above the forecast because it is the same subject one step further out — the rider
 * swipes up from the numbers, to the hours, to the sky itself.
 */
private const val VERTICAL_PAGE_RADAR = 0
private const val VERTICAL_PAGE_WEATHER = 1
private const val VERTICAL_PAGE_GAUGES = 2
private const val VERTICAL_PAGE_NAV = 3
private const val VERTICAL_PAGE_COUNT = 4

/** Gauges centre, Board Move, board Lights, diagnostics. */
private const val CONTROL_PAGE_GAUGES = 0
private const val CONTROL_PAGE_MOVE = 1
private const val CONTROL_PAGE_LIGHTS = 2
private const val CONTROL_PAGE_DIAGNOSTICS = 3
private const val CONTROL_PAGE_COUNT = 4

@Composable
private fun MirrorContent(
    state: MirrorState,
    phoneLink: PhoneLink,
    ambient: AmbientMode,
    focus: () -> Float = { 0f },
    controlFocus: () -> Float = { 0f },
    weatherFocus: () -> Float = { 0f },
    onWeatherClick: (() -> Unit)? = null,
) {
    when (state.status) {
        MirrorStatus.DISCONNECTED -> {
            if (phoneLink == PhoneLink.NO_PHONE) {
                // No arcs to pin without a phone, only the notice — and it is a readout, so it
                // leaves with them rather than bleeding under whatever page took the centre.
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .graphicsLayer { alpha = fadeOut(maxOf(focus(), controlFocus(), weatherFocus())) },
                    contentAlignment = Alignment.Center,
                ) {
                    DisconnectedLayout(ambient)
                }
            } else {
                FrameLayout(EMPTY_FRAME, false, focus, controlFocus, weatherFocus, onWeatherClick, ambient)
            }
        }
        MirrorStatus.WAITING ->
            FrameLayout(state.frame!!, false, focus, controlFocus, weatherFocus, onWeatherClick, ambient)
        MirrorStatus.STALE ->
            FrameLayout(state.frame!!, true, focus, controlFocus, weatherFocus, onWeatherClick, ambient)
        MirrorStatus.LIVE ->
            FrameLayout(state.frame!!, false, focus, controlFocus, weatherFocus, onWeatherClick, ambient)
    }
}

/** Stable gauge shell shown before the first phone frame; every board lane renders as disabled. */
private val EMPTY_FRAME = WatchFrame(
    speed = null,
    duty = null,
    battery = null,
    motorTemp = null,
    ctrlTemp = null,
    stale = false,
)

/**
 * Ambient repaint cadence. The phone drops to a 5 s push while the wrist is in ambient
 * (`WATCH_FRAME_AMBIENT_INTERVAL_MS`), so a slower tick here only ages the reading on screen without
 * saving a single radio wake. One minute — the system's own ambient callback rate — left battery and
 * temperatures up to a minute behind data the watch already had in memory.
 *
 * @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/BoardSessionController.kt `WATCH_FRAME_AMBIENT_INTERVAL_MS`
 */
private const val AMBIENT_REFRESH_INTERVAL_MS = 10_000L

File: watch/watchos/MirrorScreen.swift
import SwiftUI

/// Screen shell around the live Watch Frame. The rim gauges (``FrameLayout``) are pinned at the
/// root and drawn once; both page axes are transparent and only swap what sits in the centre, so
/// the rider never loses speed, duty, battery and temperatures by navigating away.
///
/// The page structure is Android's, axis for axis and order for order: radar and weather above the
/// gauges, the navigation focus below, and the Board controls across. Most of those pages are
/// filled by later slices of the port — what this screen owns is that the axes exist, travel in the
/// same directions, and gate their contents the same way.
///
/// **Navigation inputs.** Swipes are kept exactly as Android has them. The Digital Crown drives the
/// vertical scroll view as well: an alternative input, not a
/// replacement for touch and not a Board control (docs/watchos.md).
///
/// Ambient does not replace this tree, it settles it: the pages stay mounted and are parked on the
/// gauges, so going in and out of the Always On state never rebuilds the screen or restarts the
/// idle clock. What ambient changes is inside ``AmbientMode``.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `MirrorScreen`
struct MirrorScreen: View {
  @ObservedObject var link: PhoneLink
  /// watchOS's single always-on signal. Wear OS gets a callback with panel capabilities; here the
  /// environment value is the whole of it.
  @Environment(\.isLuminanceReduced) private var isLuminanceReduced

  /// Optional because `scrollPosition(id:)` binds an optional; it is only nil mid-flight between
  /// pages, and every read below treats that as "not on the gauges".
  @State private var vertical: VerticalPage? = .gauges
  @State private var control: ControlPage? = .gauges
  @State private var pagePositions: [Axis: Double] = [:]
  @State private var settledPositions: [Axis: Double] = [:]
  @State private var verticalPagingEnabled = true
  /// A Board Move hold in progress. Both pagers lock while it is true and the idle return is
  /// suspended: a hold must not be read as a page swipe, and the page must never move out from
  /// under a finger that is driving a motor.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `moveHeld`
  @State private var moveHeld = false
  @GestureState private var touching = false
  @GestureState private var dragging = false
  @State private var lastInteraction = Date()
  /// Foreground/background, which is what decides whether the Mirror is awake at all. Ambient is a
  /// second, narrower question asked only while it is.
  @Environment(\.scenePhase) private var scenePhase

  private var ambient: AmbientMode { AmbientMode(active: isLuminanceReduced) }

  var body: some View {
    // A timeline rather than a timer: in the Always On state the system decides how often this
    // re-evaluates, so a stopped stream ages into `disconnected` at whatever rate the watch is
    // willing to pay for, and the degradation is itself the measurement docs/watchos.md wants.
    TimelineView(.periodic(from: .now, by: refreshInterval)) { _ in
      ZStack {
        pages
        // Pinned over the pages and never inside one: the rider keeps speed, duty, battery and
        // temperatures on every page. Nothing in it takes pointer input, so the pages below stay
        // reachable through it.
        frame
          .allowsHitTesting(false)
      }
      .onPreferenceChange(PagePositionKey.self) { values in
        var transaction = Transaction()
        transaction.disablesAnimations = true
        withTransaction(transaction) {
          pagePositions = values
          settledPositions = [:]
        }
      }
      .task(id: pagePositions) {
        // A cancellable quiet period validates measured alignment, never an assumed animation
        // duration. Every movement cancels this task and closes the interaction gate immediately.
        let positions = pagePositions
        try? await Task.sleep(for: .milliseconds(100))
        guard !Task.isCancelled else { return }
        settledPositions = positions
        if let horizontal = positions[.horizontal], abs(horizontal.rounded() - horizontal) < 0.001 {
          verticalPagingEnabled = abs(horizontal) < 0.001
        }
      }
      .simultaneousGesture(
        DragGesture(minimumDistance: 0)
          .updating($touching) { _, state, _ in state = true }
          .updating($dragging) { value, state, _ in
            state = hypot(value.translation.width, value.translation.height) > 6
          }
      )
      .onChange(of: touching) { _, _ in lastInteraction = Date() }
      // A hold both suspends the countdown and, on release, restarts it: the 45 s is measured from
      // the last thing the rider did, and a long hold is very much something they did.
      .onChange(of: moveHeld) { _, _ in lastInteraction = Date() }
      .task(id: lastInteraction) {
        try? await Task.sleep(for: .seconds(CONTROL_IDLE_RETURN_SECONDS))
        guard !Task.isCancelled, !touching, !moveHeld, !isLuminanceReduced, control != .gauges
        else { return }
        withAnimation { control = .gauges }
      }
      .task(id: tick) { link.refresh() }
    }
    .onChange(of: isLuminanceReduced) { _, reduced in
      reportWakeLevel()
      // The wrist goes always-on wherever the rider left it. Ambient only ever draws the gauges, so
      // park both axes there first — otherwise the gauges would be pinned over a control page the
      // rider can no longer swipe away.
      guard reduced else { return }
      vertical = .gauges
      control = .gauges
      verticalPagingEnabled = true
    }
    .onChange(of: scenePhase) { _, _ in reportWakeLevel() }
    .onAppear { reportWakeLevel() }
  }

  /// The phone owns the push cadence but not the fact it turns on: only the wrist knows whether it
  /// is on screen, in the Always On state, or gone. Reported on every change of either input.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MainActivity.kt `wakeHeartbeat`
  private func reportWakeLevel() {
    let level: WatchMirrorWakeLevel
    switch scenePhase {
    case .active: level = isLuminanceReduced ? .ambient : .active
    // Inactive is the Always On state's phase as much as it is a transition, so it is reported as
    // ambient rather than as gone: a wrist that is still drawing wants frames, only fewer.
    case .inactive: level = .ambient
    default: level = .asleep
    }
    link.reportWakeLevel(level)
  }

  /// Re-evaluation cadence: the phone's own push interval while the screen is on, and the reduced
  /// ambient one while it is not.
  private var refreshInterval: TimeInterval {
    isLuminanceReduced ? AMBIENT_REFRESH_INTERVAL_SECONDS : ACTIVE_REFRESH_INTERVAL_SECONDS
  }

  /// Something that changes on every timeline beat, so the reduce runs once per beat instead of
  /// once per body evaluation.
  private var tick: Int { Int(Date().timeIntervalSince1970 / refreshInterval) }

  /// Whether the radar page is allowed to fetch. It is the one surface on the wrist that spends
  /// network on its own, so this is the difference between an idle watch and a fetching one: the
  /// radar page has to be the settled page, the app has to be in the foreground, and the wrist must
  /// not be in the Always On state. A wrist left on the radar page and then lowered stops fetching
  /// rather than looping on a screen nobody can see.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `radarVisible`
  private var radarVisible: Bool {
    vertical == .radar && scenePhase == .active && !isLuminanceReduced
  }

  /// The phone's forecast while it is still worth believing. Re-read per timeline beat, so an
  /// aged-out reading disappears on its own without waiting for a push that is never coming.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `freshWeather`
  private var freshWeather: WatchWeather? { link.freshWeather() }

  // MARK: - Axes

  /// One scroll view owns the whole vertical axis, and the crown with it. Android stacks the same
  /// four pages; nesting two vertical pagers there made them compete for the same drag, and the
  /// same would be true of two crown-bound views here.
  ///
  /// A paging `ScrollView` rather than `TabView(.verticalPage)`, for one reason: the vertical page
  /// style draws a dot indicator down the right edge and offers no way to turn it off —
  /// `indexDisplayMode` exists only on `.page`. Those dots land exactly on the duty gauge, which
  /// runs up that same edge. watchOS binds the Digital Crown to a scroll view natively, so this
  /// keeps the crown, keeps the swipe, gains page snapping, and lets the indicator be hidden.
  private var pages: some View {
    ScrollView(.vertical) {
      LazyVStack(spacing: 0) {
        ForEach(VerticalPage.allCases) { page in
          verticalContent(page)
            // Each page is exactly one screen, which is what makes paging land on page boundaries.
            .containerRelativeFrame([.horizontal, .vertical])
            .background {
              pagePosition(axis: .vertical, index: page.rawValue, origin: VerticalPage.gauges.rawValue)
            }
            .id(page)
        }
      }
      .scrollTargetLayout()
    }
    .coordinateSpace(name: Axis.vertical)
    .scrollTargetBehavior(.paging)
    .scrollPosition(id: $vertical)
    // The rim gauges are the furniture on this edge; a scroll bar over them is the thing being
    // fixed here, not a thing to keep.
    .scrollIndicators(.hidden)
    // Ambient has already parked the axis, and a page animation there is wasted panel.
    .scrollDisabled(isLuminanceReduced || !verticalPagingEnabled || moveHeld)
    .onChange(of: vertical) { _, _ in lastInteraction = Date() }
  }

  @ViewBuilder
  private func verticalContent(_ page: VerticalPage) -> some View {
    switch page {
    case .radar:
      RadarScreen(
        visible: radarVisible,
        forecast: freshWeather,
        riderColor: Palette.rider(link.settings.riderColor) ?? Palette.speed
      )
    case .weather:
      WeatherScreen(forecast: freshWeather, everReceived: link.weather != nil)
    case .gauges:
      controls
    case .nav:
      // Empty on purpose, like the gauges control page: this page *is* the route and the nav stack,
      // which the pinned frame already draws. It grows into the centre as the readouts leave.
      Color.clear
    }
  }

  /// The horizontal control axis, live only on the gauges page. From a control page a vertical
  /// swipe would open a blank page over something the rider is working on, which is why Android
  /// gates the vertical axis on the control page too. Each axis has its own scroll-enabled value.
  private var controls: some View {
    ScrollView(.horizontal) {
      LazyHStack(spacing: 0) {
        ForEach(ControlPage.allCases) { page in
          controlContent(page)
            .containerRelativeFrame([.horizontal, .vertical])
            .background {
              pagePosition(axis: .horizontal, index: page.rawValue, origin: ControlPage.gauges.rawValue)
            }
            .id(page)
        }
      }
      .scrollTargetLayout()
    }
    .coordinateSpace(name: Axis.horizontal)
    .scrollTargetBehavior(.paging)
    .scrollPosition(id: $control)
    .scrollIndicators(.hidden)
    // Override the outer vertical scroll lock: returning horizontally must remain possible. A held
    // Move is the one thing that closes this axis too — leaving the Move page mid-hold would cancel
    // the press and strand the rider's intent somewhere the page no longer shows.
    .environment(\.isScrollEnabled, !isLuminanceReduced && !moveHeld)
    .onChange(of: control) { _, _ in lastInteraction = Date() }
  }

  @ViewBuilder
  private func controlContent(_ page: ControlPage) -> some View {
    switch page {
    case .gauges:
      // Empty on purpose: this page *is* the gauges, pinned at the root behind this transparent
      // axis. Only the reason a stalled stream has stopped gets to use the centre.
      if case .disconnected = link.mirror.status {
        DisconnectedLayout(link: link.link, ambient: ambient)
      } else {
        WeatherReadout(
          forecast: freshWeather,
          ambient: ambient,
          // Only tappable while this page actually owns the screen; mid-transition the target
          // would swallow the drag that is moving the pager.
          onTap: interactionEnabled(.gauges) ? { withAnimation { vertical = .weather } } : nil
        )
      }
    case .move:
      MoveScreen(
        link: link,
        // The transition gate and the pager locks answer different questions: this one refuses to
        // *start* a hold on a page that has not settled, the locks stop a started hold from being
        // taken away by a swipe.
        interactionEnabled: interactionEnabled(.move),
        onHoldChanged: { moveHeld = $0 }
      )
    case .lights:
      LightsScreen(link: link, interactionEnabled: interactionEnabled(.lights))
    case .diagnostics:
      DiagnosticsPanel(link: link, interactionEnabled: interactionEnabled(.diagnostics))
    }
  }

  // MARK: - The pinned frame

  /// The gauge shell is the app, so a stalled stream keeps its gauges and reads the reason inside
  /// them. Dropping to a bare notice threw the layout away exactly when the rider was already lost,
  /// and it hid that the clock and the battery gauge still had something to say.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `MirrorContent`
  @ViewBuilder
  private var frame: some View {
    switch link.mirror.status {
    case .disconnected:
      layout(EMPTY_FRAME, muted: false, showReadouts: false)
    case .waiting, .live:
      layout(link.mirror.frame ?? EMPTY_FRAME, muted: false)
    case .stale:
      layout(link.mirror.frame ?? EMPTY_FRAME, muted: true)
    }
  }

  private func layout(_ frame: WatchFrame, muted: Bool, showReadouts: Bool = true) -> FrameLayout {
    FrameLayout(
      frame: frame,
      muted: muted,
      ambient: ambient,
      showReadouts: showReadouts,
      navFocus: navFocus,
      awayFocus: awayFocus,
      route: link.route,
      routeGeneration: link.routeGeneration,
      navColor: Palette.rider(link.settings.riderColor) ?? Palette.nav,
      navArrowEnabled: link.settings.navArrowEnabled
    )
  }

  /// Read actual page displacement so dragging, cancelling, crown scrolling and snapping all
  /// drive the same fade. Measure every mounted page because lazy stacks can unload the gauges.
  private func pagePosition(axis: Axis, index: Int, origin: Int) -> some View {
    GeometryReader { geometry in
      let bounds = geometry.frame(in: .named(axis))
      let length = axis == .horizontal ? geometry.size.width : geometry.size.height
      let offset = axis == .horizontal ? bounds.minX : bounds.minY
      let position = length > 0 ? Double(index - origin) - Double(offset / length) : 0
      Color.clear.preference(
        key: PagePositionKey.self,
        value: [axis: position]
      )
    }
  }

  /// Where the vertical axis is, relative to the gauges: positive toward navigation, negative
  /// toward weather and radar.
  private var verticalPosition: Double {
    pagePositions[.vertical] ?? Double((vertical ?? .gauges).rawValue - VerticalPage.gauges.rawValue)
  }

  private var horizontalPosition: Double {
    pagePositions[.horizontal] ?? Double((control ?? .gauges).rawValue)
  }

  /// The nav page taking over. The readouts leave for it; the nav stack is what it is made of, so
  /// that stays and grows.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `navFocus`
  private var navFocus: Double {
    guard !isLuminanceReduced else { return 0 }
    return min(1, max(0, verticalPosition))
  }

  /// Any other page taking over — the control axis, weather, radar. Those want the whole centre, so
  /// the nav stack leaves with the readouts.
  private var awayFocus: Double {
    guard !isLuminanceReduced else { return 0 }
    return min(1, max(max(0, -verticalPosition), abs(horizontalPosition)))
  }

  // MARK: - Transition gating

  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `activePage`
  private func interactionEnabled(_ page: ControlPage) -> Bool {
    guard !isLuminanceReduced, !dragging, control == page, vertical == .gauges,
      let horizontal = settledPositions[.horizontal],
      let verticalPosition = settledPositions[.vertical]
    else { return false }
    return abs(horizontal - Double(page.rawValue)) < 0.001 && abs(verticalPosition) < 0.001
  }

}

/// Radar and weather above the gauges, navigation focus below. Radar sits above the forecast
/// because it is the same subject one step further out — the rider swipes up from the numbers, to
/// the hours, to the sky itself.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `VERTICAL_PAGE_RADAR`
enum VerticalPage: Int, CaseIterable, Identifiable {
  case radar
  case weather
  case gauges
  case nav

  var id: Int { rawValue }
}

/// Gauges centre, Board Move, board Lights, diagnostics.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `CONTROL_PAGE_GAUGES`
enum ControlPage: Int, CaseIterable, Identifiable {
  case gauges
  case move
  case lights
  case diagnostics

  var id: Int { rawValue }
}

/// Idle time after which the horizontal axis drifts back to the gauges.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorScreen.kt `CONTROL_IDLE_RETURN_MS`
private let CONTROL_IDLE_RETURN_SECONDS: TimeInterval = 45

/// Each page on an axis reports the same displacement, including when the gauges are offscreen.
private struct PagePositionKey: PreferenceKey {
  static let defaultValue: [Axis: Double] = [:]

  static func reduce(value: inout [Axis: Double], nextValue: () -> [Axis: Double]) {
    value.merge(nextValue(), uniquingKeysWith: { current, _ in current })
  }
}
